planner/util/path.go

```diff
@@ -85,7 +85,17 @@ func (path *AccessPath) SplitCorColAccessCondFromFilters(ctx sessionctx.Context,
 	for i := eqOrInCount; i < len(path.IdxCols); i++ {
 		matched := false
 		for j, filter := range path.TableFilters {
-			if used[j] || !isColEqCorColOrConstant(ctx, filter, path.IdxCols[i]) {
+			if used[j] {
+				continue
+			}
+			colEqConstant := isColEqConstant(filter, path.IdxCols[i])
+			if i == eqOrInCount && colEqConstant {
+				// If there is a col-eq-constant condition for path.IdxCols[eqOrInCount], it means that range fallback happens
```

Comment:
I can't understand this. Why does a `col = constant` condition mean fallback happened?

Reply:
If range fallback doesn't happen, `DetachCondAndBuildRangeForIndex` must put `col = constant` to `path.AccessConds` rather than `path.TableFilters`.

Reply:
`i == eqOrInCount && colEqConstant` IF is to prevent the following case:
```
create table t (a int, b int, c int, index idx_a_b(a, b));
set @@tidb_opt_range_max_size=1000;
explain format='brief' select /*+ use_index(t, idx_a_b) */ * from t where a in (1, 3, 5) and b = 2;
```
`DetachCondAndBuildRangeForIndex` puts `a in (1, 3, 5)` into `path.AccessConds` and puts `b = 2` into `path.TableFilters`. Without `i == eqOrInCount && colEqConstant` IF, `SplitCorColAccessCondFromFilters` adds `b = 2` back into `path.AccessConds` but doesn't rebuild ranges, which leads to wrong results.

Reply:
Yes. The IF branch ensures that there must be some `col-eq-corcol` condition in `access` if `len(access) > 0`. Actually the first one in `access` must be  some `col-eq-corcol` condition. I add it to comment in code.